cristin-organization/src/main/java/no/unit/nva/cristin/organization/CristinApiClient.java

```diff
@@ -45,15 +52,65 @@ public Organization getOrganization(URI uri)
-                        .addChild(UriUtils.INSTITUTION)
-                        .addQueryParameters(requestQueryParams)
-                        .getUri())
-                .withProcessingTime(0L)
-                .withSize(0)
-                .withHits(Collections.emptyList());
+    public SearchResponse<Organization> queryOrganizations(Map<String, String> requestQueryParams)
+            throws NotFoundException, FailedHttpRequestException {
+        Map<String, String> cristinRequestQueryParams = translateToCristinApi(requestQueryParams);
+        URI queryUri = new UriWrapper(HTTPS, CRISTIN_API_BASE)
+                .addChild(UNITS_PATH)
+                .addQueryParameters(cristinRequestQueryParams)
+                .getUri();
+
+        long start = System.currentTimeMillis();
+        SearchResponse<Organization> searchResponse = httpExecutor.query(queryUri);
+
+        return updateSearchResponseMetadata(searchResponse,requestQueryParams, System.currentTimeMillis() - start);
+    }
+
+    private Map<String, String> translateToCristinApi(Map<String, String> requestQueryParams) {
+        return Map.of(
+                CRISTIN_QUERY_NAME_PARAM, requestQueryParams.get(QUERY),
+                PAGE, requestQueryParams.get(PAGE),
+                CRISTIN_PER_PAGE_PARAM, requestQueryParams.get(NUMBER_OF_RESULTS));
     }
+
+    private SearchResponse<Organization> updateSearchResponseMetadata(
+            SearchResponse<Organization> searchResponse,
+            Map<String, String> requestQueryParams,
+            long timeUsed) {
+        final URI baseUri = new UriWrapper(HTTPS, DOMAIN_NAME).addChild(BASE_PATH).addChild(ORGANIZATION_PATH).getUri();
+        searchResponse.setId(new UriWrapper(baseUri).addQueryParameters(requestQueryParams).getUri());
+        searchResponse.setFirstRecord(calculateFirstRecord(requestQueryParams));
+        searchResponse.setNextResults(nextResult(baseUri, requestQueryParams, searchResponse.getSize()));
+        searchResponse.setPreviousResults(previousResult(baseUri, requestQueryParams, searchResponse.getSize()));
+        searchResponse.setProcessingTime(timeUsed);
+        return searchResponse;
+    }
+
+    private URI previousResult(URI baseUri, Map<String, String> requestQueryParams, int totalSize) {
+        int firstPage = Integer.parseInt(requestQueryParams.get(PAGE)) - 1;
+        if (firstPage > 0 && totalSize > 0) {
+            Map<String, String> nextMap = new ConcurrentHashMap<>(requestQueryParams);
+            nextMap.put(PAGE, Integer.toString(firstPage));
+            return new UriWrapper(baseUri).addQueryParameters(nextMap).getUri();
+        }
+        return null;
+    }
+
+    private URI nextResult(URI baseUri, Map<String, String> requestQueryParams, int totalSize) {
+        int nextPage = Integer.parseInt(requestQueryParams.get(PAGE)) + 1;
+        int pageSize = Integer.parseInt(requestQueryParams.get(NUMBER_OF_RESULTS));
+        if (nextPage * pageSize < totalSize) {
```

Comment:
This line works only if the total number of results is an exact multiple of pageSize. if totalSize == 29 and pageSize==10 then nextPage*pageSize == 30 and the last 9 results will not be visible.

Reply:
There is logic for this in `SearchResponse` but it depends on headers from Cristin response

Reply:
Recalculating nextResult in another/better way...   At this point we cannot use the cristin headers as we rewrite the uri